### PR TITLE
Add ability to import/export tagging rules

### LINKED ISSUE #2284: Export and Import Tagging rules
### Issue details

I would love to see a feature, where i can export my tagging rules since i have more than 10 now. It should act as backup, so when i move to a different host i can export my articles and my tagging rules from the old host and can import both exported files on the new host and be up and running again.
(bonus would be if it is in one file (articles+tagging rules)

### PR DESCRIPTION
**Pull request by j0k3r**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | dunno
| Documentation | no
| Translation   | yes
| License       | MIT

Fix #2284

image

- [x] Add baggy theme
- [x] Add all missing translations
- [x] Add some tests
- [x] Add API

The API will be accessible at `/api/taggingrule/export` it'll return the exported rules as a json file.

### CODE REVIEW

File: src/Wallabag/ApiBundle/Controller/TaggingRuleRestController.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+            'json',
+            SerializationContext::create()->setGroups(['export_tagging_rule'])
+        );
+
+        return Response::create(
+            $data,
+            200,
+            [
+                'Content-type' => 'application/json',
+                'Content-Disposition' => 'attachment; filename="tagging_rule_' . $this->getUser()->getUsername() . '.json"',
```
Comment (tcitworld) [inline src/Wallabag/ApiBundle/Controller/TaggingRuleRestController.php]:
```suggestion
                'Content-Disposition' => 'attachment; filename="tagging_rules_' . $this->getUser()->getUsername() . '.json"',
```
We could also add the date.

File: src/Wallabag/CoreBundle/Controller/ConfigController.php [hunk 2]
```diff
@@ -140,13 +144,45 @@ public function indexAction(Request $request)
...
+
+                    $em->flush();
+
+                    $this->addFlash(
+                        'notice',
+                        'flashes.config.notice.tagging_rules_imported'
+                    );
+                }
+            }
+
```
Comment (tcitworld) [inline src/Wallabag/CoreBundle/Controller/ConfigController.php]: Maybe add an error message before returning ?

File: src/Wallabag/CoreBundle/Controller/ConfigController.php [hunk 3]
```diff
@@ -492,6 +528,32 @@ public function setLocaleAction(Request $request, $language = null)
...
+            'json',
+            SerializationContext::create()->setGroups(['export_tagging_rule'])
+        );
+
+        return Response::create(
+            $data,
+            200,
+            [
+                'Content-type' => 'application/json',
+                'Content-Disposition' => 'attachment; filename="tagging_rule_' . $this->getUser()->getUsername() . '.json"',
```
Comment (tcitworld) [inline src/Wallabag/CoreBundle/Controller/ConfigController.php]:
```suggestion
                'Content-Disposition' => 'attachment; filename="tagging_rules_' . $this->getUser()->getUsername() . '.json"',
```
Same, maybe add date.

File: tests/Wallabag/CoreBundle/Controller/ConfigControllerTest.php [hunk 4]
```diff
@@ -1097,4 +1098,67 @@ public function testUserEnable2faGoogleCancel()
...
+
+    public function testImportTagginfRuleFile()
+    {
+        $this->logInAs('admin');
+        $client = $this->getClient();
+
+        $crawler = $client->request('GET', '/config');
+        $form = $crawler->filter('form[name=upload_tagging_rule_file] > button[type=submit]')->form();
+
+        $file = new UploadedFile(__DIR__ . '/../fixtures/tagging_rule_admin.json', 'tagging_rule_admin.json');
```
Comment (tcitworld) [inline tests/Wallabag/CoreBundle/Controller/ConfigControllerTest.php]:
```suggestion
        $file = new UploadedFile(__DIR__ . '/../fixtures/tagging_rules_admin.json', 'tagging_rule_admin.json');
```

File: tests/Wallabag/CoreBundle/Controller/ConfigControllerTest.php [hunk 5]
```diff
@@ -1097,4 +1098,67 @@ public function testUserEnable2faGoogleCancel()
...
+
+        ob_start();
+        $crawler = $client->request('GET', '/tagging-rule/export');
+        ob_end_clean();
+
+        $this->assertSame(200, $client->getResponse()->getStatusCode());
+
+        $headers = $client->getResponse()->headers;
+        $this->assertSame('application/json', $headers->get('content-type'));
+        $this->assertSame('attachment; filename="tagging_rule_admin.json"', $headers->get('content-disposition'));
```
Comment (tcitworld) [inline tests/Wallabag/CoreBundle/Controller/ConfigControllerTest.php]:
```suggestion
        $this->assertSame('attachment; filename="tagging_rules_admin.json"', $headers->get('content-disposition'));
```

**Comment by j0k3r** [comment]
Squashed commits

**Comment by tcitworld** [comment]
https://github.com/wallabag/doc/issues/105

**Pull request merged**